lib/vast/Conversion/ABI/EmitABI.cpp

```diff
@@ -134,21 +143,31 @@ namespace vast
                 VAST_ASSERT(a_it == a.end() && b_it == b.end());
             }
 
-            void zip_ret(const auto &a, const auto &b, auto &&yield)
+            void zip_ret(const auto &abi, const auto &results, const auto &concretes,
+                         auto &&yield)
             {
-                auto abi_it = a.begin();
-                auto args_it = b.begin();
+                // These should be always synced, but we are defensive as this may
+                // be called with some random module that is not a result of our
+                // classification.
+                VAST_ASSERT(std::ranges::size(abi) == std::ranges::size(results));
+                auto abi_it = abi.begin();
```

Comment:
what about `llvm::zip`, probably can be applied also in the previous `zip` function

Reply:
Will add comment, but this probably will not be classic `zip` in the end once we start supporting other `abi` classifications (`sret` for example).